Language selection interstitial when first opening the app

akhavich opened the pull request:
# Pull Request

## Description

Adds an interstitial page allowing the user to select their preferred language when it isn't saved in the browser's local storage or in UserSettings in the database. In practice, this should only popup the first time a user accesses the app in any given browser (or private window). It saves the chosen language in LocalStorage and will also persist it to UserSettings when the user is logged in.

![image](https://github.com/user-attachments/assets/e525fc0f-8b8d-45ce-a775-ef05b2692db5)


## Related Issues
Task # 6960 - Add culture to the local storage
Task # 6962 - On first load, check local storage for culture and update if found
Task # 6963 - Redirect to a language page if culture not found
Task # 6964 - Create a language page to set culture

## Changes
<!-- List the key changes in this PR -->

- Added a language selection page accessible by non-logged-in users
- Ensure that the selected language from that page is always saved to browser local storage
- Keep local storage and UserSettings values in sync

## Testing
<!-- Describe the tests that were run or any QA steps that were taken -->

- Manual testing with browser session logged in, logged out, with and without UserSettings and local storage set.

## Checklist
<!-- Ensure the following tasks are complete -->

- [x] Code follows dotnet coding standards
- [ ] Tests added/updated to cover changes

akhavich committed:
Convert CultureService to interface/impl model

akhavich committed:
Language selection page with redirect to previous url

akhavich committed:
Added OverrideCulture to allow showing T&C in either language without persisting it

akhavich committed:
Persist language from localstorage to UserSettings

akhavich committed:
Save language preference from UserSettings in localstorage if localstorage doesn't have it

akhavich committed:
Update user language setting after login with different language preference

akhavich committed:
Merge remote-tracking branch 'origin/develop' into ak/language_selection

Sean-Stilwell commented on Portal/src/Datahub.Portal/Pages/Public/LanguageSelectionPage.razor:
```diff
@@ -0,0 +1,59 @@
+            <MudStack>
+                <MudImage Src="/img/canada-fip.svg" Alt="Government of Canada Wordmark / Symbole de mot du gouvernement du Canada" Style="max-height: 2rem" />
+                <MudSpacer />
+                <MudStack Row>
```
We should include the FSDH + DHSF titles in this dialog so the user knows where they are.

As well as centre the buttons as John mentioned.

Reference for how FOSRC handles it:

![image](https://github.com/user-attachments/assets/213cb7ee-c6c6-4a54-b4c5-1cd3a1a4445d)

yjmrobert replied:
If we could also get the background color on the footer that would be 💯. Will probably have to rework the padding down into the pieces of the `MudPaper`

Sean-Stilwell commented on Portal/src/Datahub.Portal/Layout/BarebonesPublicLayout.razor:
```diff
@@ -0,0 +1,26 @@
+
+@code {
+
+    private readonly string _backgroundStyle = new StyleBuilder()
```
Low priority, can we blur this to match other GC sites? It would be good for consistency.

Ex: https://open-science.canada.ca/

yjmrobert replied:
Total tangent but it's gross how they load every image and programmatically add the `hide` class instead of just programmatically changing the source url...

![image](https://github.com/user-attachments/assets/809e4660-38f9-4068-8444-ea630bae0ed0)

yjmrobert replied:
@gcalexk if you want you can just add something like a `filter: blur(4px)`

Sean-Stilwell commented on Portal/src/Datahub.Portal/Pages/Public/LanguageSelectionPage.razor:
```diff
@@ -0,0 +1,59 @@
+                        Avis
+                    </MudLink>
+                    <MudSpacer />
+                    <MudImage Src="/img/fip_mark.png" Alt="Canada" Style="max-height: 1rem" />
```
If we can, I'd boost the size of this, it's super tiny on my display.

Sean-Stilwell reviewed (changes requested):
Awesome update! The language selection page is definitely needed for official languages.

I have some recommended changes to the UI. I'd also like to see if @davidreneuw has any a11y points.

I think it also needs unit tests

yjmrobert commented on Portal/src/Datahub.Portal/_Imports.razor:
```diff
@@ -69,4 +70,4 @@
 @using PSC.Blazor.Components.MarkdownEditor.EventsArgs
 
 @inject IStringLocalizer Localizer
-@inject CultureService CultureService
```
Argh I didn't know this was globally injected into every component. Probably best practice to remove it - unless it's a lot of work and there's a ton of affected components.

Thoughts?  (feel free to disregard)

yjmrobert reviewed (approved):
This is great! 

A few subtle things that Sean pointed out around the GC specs but those can be iterated on at a later date if you're tight on cycles (please add them to the boards if that's the case!)

akhavich committed:
Removed global ICultureService injection; put it in individual razor components/pages

akhavich committed:
Blur background image

akhavich committed:
Centered buttons, added FSDH/DHSF text, enlarged Canada logo

akhavich committed:
Merge remote-tracking branch 'origin/develop' into ak/language_selection

akhavich commented:
Latest updates:
- Blurred background - was a bit more involved than just filter:blur and I'm still not quite happy with it (there's a white border) but getting rid of that turns out to be a fiddly rabbit hole. Maybe address in the future
- Removed ICultureService injection from _Imports and added back to the individual razor pages/components
- Centered buttons and added FSDH/DHSF text
- Enlarged Canada logo (along with T&C/Avis text and the whole dialog width)

Still outstanding:
- Unit tests: Do we have the capability to unit test browser-based features? If not, there's not much we can do with localstorage
- Background colour on the footer
- Visual tweaks

![image](https://github.com/user-attachments/assets/53b32820-b1a6-41bd-ad2c-0e03f51abb8c)

yjmrobert reviewed (approved):
Great stuff, I will add the outstanding tasks to the boards

Sean-Stilwell reviewed (approved):
Awesome stuff.

Funny how it lines up to spell the title across as well as vertically.

![image](https://github.com/user-attachments/assets/9b08acc4-2336-460d-89d4-98a55c3d5e26)

davidreneuw committed:
🔧 Accessibility changes

davidreneuw commented:
@akhavich I have made a few small changes to ensure all accessibility requirements are met:

- Fixed heading levels by adding a hidden H1 at the top of the page and replacing the H3s with H2s
- Added lang attributes to the different sections to specifically define the presence of different languages
- Added translate="no" to stop elements from being automatically translated by browsers

Normally, these changes should have no visual impact on this page. These changes were all made in direct reflection of canada.ca
